tests/phpunit/includes/build-equivalent-html-semantic-tree.php

```diff
@@ -0,0 +1,299 @@
+<?php
+
+/* phpcs:disable WordPress.Security.EscapeOutput.ExceptionNotEscaped */
+
+/**
+ * Generates the tree-like structure represented in the Html5lib tests.
+ *
+ * That format is extended with a special representation of block delimiters and their attributes.
+ * Furthermore, it the order of attributes and class names is normalized both for HTML tags and block delimiters,
+ * as is the whitespace in HTML tags' style attribute.
+ *
+ * For example, consider the following block markup:
+ *
+ *     <!-- wp:separator {"className":"is-style-default has-custom-classname","style":{"spacing":{"margin":{"top":"50px","bottom":"50px"}}},"backgroundColor":"accent-1"} -->
+ *         <hr class="wp-block-separator is-style-default has-custom-classname" style="margin-top: 50px; margin-bottom: 50px" />
+ *     <!-- /wp:separator -->
+ *
+ * This will be represented as:
+ *
+ *     BLOCK["core/separator"]
+ *       {
+ *         "backgroundColor": "accent-1",
+ *         "className": "has-custom-classname is-style-default",
+ *         "style": {
+ *           "spacing": {
+ *             "margin": {
+ *               "top": "50px",
+ *               "bottom": "50px"
+ *             }
+ *           }
+ *         }
+ *       }
+ *       <hr>
+ *         class="has-custom-classname is-style-default wp-block-separator"
+ *         style="margin-top:50px;margin-bottom:50px;"
+ *
+ *
+ * @see https://github.com/WordPress/wordpress-develop/blob/trunk/tests/phpunit/data/html5lib-tests/tree-construction/README.md
+ *
+ * @param string|null $fragment_context Context element in which to parse HTML, such as BODY or SVG.
+ * @param string      $html             Given test HTML.
+ * @return string|null Tree structure of parsed HTML, if supported, else null.
+ */
+function build_equivalent_html_semantic_tree( string $html, ?string $fragment_context ) {
```

Comment:
There's no equivalence at this point. It's a single tree representation of the HTML (and block) semantics. Equivalence only enters the picture in the assertion where we compare two trees to understand whether they're semantically equivalent.

Let's remove the notion of equivalence from this function name:

```suggestion
function build_html_semantic_tree( string $html, ?string $fragment_context ) {
```

Reply:
while we’re at it, it seems like the _semantic_ part is stored inside the HTML API and not in the output text. is this actually more a _visual_ tree?

```php
function build_visual_html_tree() {}
```

I ask because if we wanted to compare apples to apples I think we can just normalize two documents, or provide two HTML Processors, but I think a reasonable part of this function is providing better error display when the HTML is not equivalent?

Reply:
That's a good point, I've struggled to figure out the language myself. It's a _visual_ or human-readable representation of the tree. I landed on semantic because, although it's imperfect, it conveys that it represents the semantics of the HTML rather than the syntax.

It may be the case that "tree" already suggests that it's a processed form of the HTML input.

I'm happy with "visual," more details about what the function does are available in the PHP doc comment.

Reply:
Alright, I'll rename to `build_visual_html_tree`!

Reply:
9469f43d0bd85ffbc7c6ea24f9db3d724d6a9f3f